pyro/poutine/posterior_messenger.py

```diff
@@ -0,0 +1,11 @@
+from .replay_messenger import ReplayMessenger
+
+
+class SamplePosteriorMessenger(ReplayMessenger):
+    # This acts like ReplayMessenger but additionally replays cond_indep_stack.
+
+    def _pyro_sample(self, msg):
+        if msg["infer"].get("enumerate") == "parallel":
+            super(SamplePosteriorMessenger, self)._pyro_sample(msg)
+        if msg["name"] in self.trace:
+            msg["cond_indep_stack"] = self.trace.nodes[msg["name"]]["cond_indep_stack"]
```

Comment:
You will also need to call the parent method towards the end, otherwise you wouldn't be replaying, and just sampling from the prior! That should fix the test, I suppose?

```python
super(SamplePosteriorMessenger, self)._pyro_sample(msg)
```

It will be great if you added some tests to `test_abstract_infer`! Existing tests are pretty bare and only exercise MCMC, and it will be good to have some tests for `SVI` too.

Reply:
Thanks!

Reply:
I will try to add some more tests 😄

Reply:
I am trying to add a test with the following guide, but I am running into shape issues. Any ideas what could be wrong? Thanks! 😄 

```python
def guide(num_trials):
  phi_c0 = pyro.param("phi_c0", num_trials.new_ones(num_trials.shape[0]) * 500.0)
  phi_c1 = pyro.param("phi_c1", num_trials.new_ones(num_trials.shape[0]) * 500.0)
  phi_posterior = dist.Beta(concentrantion0=phi_c0, concentrantion1=phi_c1)
  pyro.sample("phi", phi_posterior)
```

If I have it like above, I get an error about invalid shape of `log_prob` and that I should try using `to_event`. If I add `to_event(1)` on the `phi_posterior` I get an error about a mismatch between the shape in the model and the one in the guide. This also seems to happen when I use an automatic guide like `AutoDelta`

Reply:
I would suggest using a plate over the data dims in both model and guide:

```python
...
with pyro.plate("num_trials", num_trials):
    pyro.sample("phi", phi_posterior)
```

Feel free to commit your test, and we can take another look, if that doesn't work.

Reply:
Putting it in a plate seems to solve the issue I mentioned!
Unfortunately, it seems that with the manual guide, it look like it is sampling from the prior again? Maybe I misspecified something, but I thought that `Beta` was conjugate to `Binomial` and so the guide should work well.

Reply:
The auto guide tests, work however!

Reply:
One issue is that with `Beta(500, 500)`, you are placing a heavy prior around 0.5. Consider using a weaker prior and higher learning rate.

Reply:
Thanks for the comments! Relaxing the parameter values helped.
Funnily I was actually aiming to select parameter values of the posterior in the guide close to the actual values of the true posterior `300` and `700` I think?